### PR TITLE
Merkle distributor contracts (optional/mandatory)

### PR DESCRIPTION
**Pull request by tomlinton**
Draft of the optional and mandatory lockup distributor contracts. @sparrowDom @franckc said you might be interested in working on these. We need to add a shutdown state to these contracts where we can reclaim any OGV on the distributor contracts after the claim period has expired.

Edit (Domen):
- added shutdown state (burning remaining OGV) once the claim period expires. 

**Related issue:** https://github.com/OriginProtocol/ousd-governance/issues/109

### CODE REVIEW

**Comment by tomhirst** [comment]
FYI @tomlinton added a fix for the deploy script, so I can merge this into #82 and start integrating the contracts with the UI. Hope that's OK!

**Comment by tomlinton** [comment]
Yep no problem, I think I broke it with a bad merge

File: contracts/MandatoryLockupDistributor.sol
Comment (DanielVF) [inline contracts/MandatoryLockupDistributor.sol:41]: @tomlinton What's the purpose of this index bitmasking dance vs just reading and writing a boolean to a mapping?

**Comment by DanielVF** [comment]
Looks like these are missing the mentioned OGV recovery methods after shutdown?

File: contracts/MandatoryLockupDistributor.sol
Comment (tomlinton) [reply contracts/MandatoryLockupDistributor.sol:41]: It's a fork of the uniswap merkle distributor which seems easier than rewriting it. I'd guess it's also more gas efficient.

**Comment by tomlinton** [comment]
That can be separate PR, for now I just want to merge this so we can get a Rinkeby setup going.

**Comment by sparrowDom** [comment]
@tomlinton feel free to merge this. Have added a burnable function but have not tested it. Will do any fixes in subsequent PRs.

File: contracts/GovernanceToken.sol
Comment (DanielVF) [inline contracts/GovernanceToken.sol:4]: I see the change to the import, but does this actually get used?

Comment (sparrowDom) [reply contracts/GovernanceToken.sol:4]: My bad forgot to add it to contract definition. Should be ok now

File: contracts/MandatoryLockupDistributor.sol
Comment (DanielVF) [inline contracts/MandatoryLockupDistributor.sol:28]: Spelling: "eligible"

**Comment by DanielVF** [comment]
Are tests passing for you guys? 

I'm getting a lot of failures: 

image

Looks like most are around making a proposal.

image

**Comment by tomlinton** [comment]
Needs --network hardhat because it uses hardhat_mine calls

File: contracts/MandatoryLockupDistributor.sol
Comment (DanielVF) [inline contracts/MandatoryLockupDistributor.sol:36]: Just checking a design decision here. These claim methods don't stop working after the endBlock, it's currently more a soft end until we burn the funds and the staking calls break on balance transfer errors.


Do we want to just have a `require(block.number < endBlock)` in these claim methods, so the end requires no human action, and is explicit and precise?

Comment (sparrowDom) [reply contracts/MandatoryLockupDistributor.sol:36]: Sounds like a good proposal to me. Adding that in!

**Comment by sparrowDom** [comment]
@DanielVF change done

File: tests/helpers.py
Comment (DanielVF) [inline tests/helpers.py:13]: Did we mean to have these two prints in here?

File: tests/helpers.py
Comment (DanielVF) [inline tests/helpers.py:15]: Notice that this method of computing differences will crash if both the actual and the expected are both zero. ( `a+b` will equal zero, which will cause a divide by zero error.)

This math does seem to work for small differences, > 3 decimals, but it gets weird for low decimal precision. The difference between 50 and 100 comes out to 0.66666, which probably isn't what you would expect from such a large difference.

File: tests/helpers.py
Comment (sparrowDom) [reply tests/helpers.py:13]: not sure why those are there... tnx removing

File: tests/helpers.py
Comment (sparrowDom) [reply tests/helpers.py]: good feedback thanks. have changed the function to: 
- always return false if expected / received numbers have different sign
- instead of a / b have renamed properties as `expected`and `actual`. This way the denominate using the `exptected` property and not the average between expected and actual. so 50 (expected) and 100(actual) will yield `2` as in 200%. Which is more correct IMHO

**Comment by sparrowDom** [comment]
@DanielVF can we merge this in?

**Review by DanielVF — APPROVED** [review]
These contracts look good. Couple very minor things noted, but after those are fixed, we are good to merge.

**Comment by tomlinton** [comment]
Added those final things thanks @DanielVF, will merge when tests pass.

**Comment by DanielVF** [comment]
Rock On!

**Pull request merged**
